lucene/expressions/build.gradle

```diff
@@ -32,8 +32,8 @@ dependencies {
   // Should we really go through these hoops?
   implementation 'org.ow2.asm:asm'
   implementation('org.ow2.asm:asm-commons', {
-    exclude group: "org.ow2.asm", module: "asm-tree"
```

Comment:
I would like to understand why this needs to be done.

Reply:
It seems there are a kind of transitive dependencies (I'm not sure this terminology is correct) from `asm-commons` to `ams-tree` and `asm-analysis`. This propagation could be stopped, I haven't examined yet; this is not the main topic of the PR and I thought it can be delayed.
```
lucene $ jar --describe-module --file ~/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-commons/7.2/ca2954e8d92a05bacc28ff465b25c70e0f512497/asm-commons-7.2.jar 
org.objectweb.asm.commons@7.2.0 jar:file:///home/moco/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-commons/7.2/ca2954e8d92a05bacc28ff465b25c70e0f512497/asm-commons-7.2.jar/!module-info.class open
exports org.objectweb.asm.commons
requires java.base mandated
requires org.objectweb.asm transitive
requires org.objectweb.asm.tree transitive
requires org.objectweb.asm.tree.analysis transitive
```

Reply:
javac complains like this with the `exclude` statements.
```
lucene $ ./gradlew -p lucene/expressions/ compileJava -Duser.language=en
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details

> Task :errorProneSkipped
WARNING: errorprone disabled (skipped on non-nightly runs)

> Task :lucene:expressions:compileJava FAILED
error: module not found: org.objectweb.asm.tree.analysis
error: module not found: org.objectweb.asm.tree
2 errors
```

Reply:
(Even if the transitive dependency can be stopped (as before), intentionally breaking the dependency graph seems not good to me...)

Reply:
Thanks Tomoko, this helps. Indeed the problem is caused by the fact that the asm-commons module is a proper (named) module and has non-optional dependencies to other modules. This will indeed cause the module layer to complain. It's an interesting observation in general (for us) that if we switch to full modules we won't be able to "omit" certain dependencies of those modules anymore - the was it was possible in classpath-mode.

Reply:
>  the asm-commons module is a proper (named) module and has non-optional dependencies to other modules.

yes, I found it has a proper module-info.class. This is automatically generated and it seems to directly assemble the class file by analyzing all class files, bypassing "module-info.java"...
https://gitlab.ow2.org/asm/asm/-/blob/master/tools/bnd-module-plugin/src/main/java/org/objectweb/asm/tools/ModuleInfoBndPlugin.java

```
asm $ find asm-commons/ -name "module-info.java"
// no results
```

Reply:
Wah that's horrible!

Getting rid of asm-commons is hard, as GeneratorAdapter is very useful (the only class we use). We may fork it to expressions module.

Reply:
Isn't it Rémi Forax's project? Maybe you could ask him to relax that hard module requirement?